### PR TITLE
New prediction runtime C API, to support multiple data types (Part of #196)

### PR DESCRIPTION
**Pull request by hcho3**
Extracted from #196 for easier reviewing.
(The CI tests in this PR won't pass due to missing parts. See #196 to look up the test results.)

* The runtime now queries the data type for the model it loads, via `QueryThresholdType()` and `QueryLeafOutputType()`. Every compiled model now embeds the type information.
* Implement a full-fledged data matrix class `DMatrix` in the runtime, to replace `DenseBatch` and `SparseBatch`. The former `*Batch` classes assumed float32 data, whereas the new `DMatrix` class is able to handle both float32 and float64.
* Some API functions like `PredictBatch()` now takes `void*` pointers to accommodate multiple data types.

@higumachan Here is the pull request I would like you to review. Pay particular attention to the two C API header files, `include/treelite/c_api_common.h` and `include/treelite/c_api_runtime.h`, as they expose the API interface you will interact with in the future. Other parts of this PR are implementations of the new interface and are not part of the public-facing API.

### CODE REVIEW

File: include/treelite/c_api_runtime.h [hunk 1]
```diff
@@ -96,75 +41,37 @@ TREELITE_DLL int TreeliteBatchGetDimension(void* handle,
...
- * \return 0 for success, -1 for failure
- */
-TREELITE_DLL int TreelitePredictorPredictInst(PredictorHandle handle,
-                                              union TreelitePredictorEntry* inst,
-                                              int pred_margin, float* out_result,
-                                              size_t* out_result_size);
-
+TREELITE_DLL int TreelitePredictorPredictBatch(
+    PredictorHandle handle, DMatrixHandle batch, int verbose, int pred_margin,
+    void* out_result, size_t* out_result_size);
```
Comment (higumachan) [inline include/treelite/c_api_runtime.h]: Wouldn't we use "PredictorOutputHandle"(type of out_result) here?

File: include/treelite/c_api_runtime.h [hunk 2]
```diff
@@ -96,75 +41,37 @@ TREELITE_DLL int TreeliteBatchGetDimension(void* handle,
...
+ * \param batch the data matrix containing a batch of rows
  * \param verbose whether to produce extra messages
  * \param pred_margin whether to produce raw margin scores instead of
  *                    transformed probabilities
- * \param out_result resulting output vector; use
- *                   TreelitePredictorQueryResultSize() to allocate sufficient
- *                   space
+ * \param out_result Resulting output vector. This pointer must point to an array of length
+ *                   TreelitePredictorQueryResultSize() and of type
+ *                   TreelitePredictorQueryLeafOutputType().
```
Comment (higumachan) [inline include/treelite/c_api_runtime.h:69]: I feel it a bit confusing what values to pass to "out_result" (compared to the previous API, which was just a float).
My idea is to add a function in the API that allocates an area for out_result.
e.g. 
```c
int TreeliteCreatePredictorOutputArea(PredictorHandle handle, PredictorOutputHandle* out_result)
```

**Review by higumachan — COMMENTED** [review]
I've made a few comments.
Thanks.

File: include/treelite/c_api_runtime.h (hunk 1 above)
Comment (hcho3) [reply include/treelite/c_api_runtime.h]: The `out_result_size` is actually `LeafOutputType*`, where the type `LeafOutputType` is determined at runtime (by calling `QueryLeafOutputType()`).

Yes, let's use `PredictorOutputHandle` here to clarify the intention of `out_result_size`.

File: include/treelite/c_api_runtime.h (hunk 2 above)
Comment (hcho3) [reply include/treelite/c_api_runtime.h:69]: Good suggestion. I'll add it.

File: include/treelite/c_api_runtime.h (hunk 1 above)
Comment (hcho3) [reply include/treelite/c_api_runtime.h]: Done.

File: include/treelite/c_api_runtime.h (hunk 2 above)
Comment (hcho3) [reply include/treelite/c_api_runtime.h:69]: I added C API functions `TreeliteCreatePredictorOutputVector()` and `TreeliteDeletePredictorOutputVector()`. I also added detailed instructions on accessing element values from the output vector.

**Comment by hcho3** [comment]
@higumachan Can you take a second look?

**Comment by higumachan** [comment]
@hcho3 
Looks good to me!
Thanks.

**Comment by hcho3** [comment]
@higumachan Thanks! The refactor will be complete once #198 and #201 are reviewed. I have asked my colleagues to review them. I'll let you know once all the refactoring changes make it to the mainline branch.

**Comment by higumachan** [comment]
@hcho3 Okay, I am waiting your announce.
Am I okay with starting to build an API for Rust based on this PR C API yet?(I'm probably going to have to build it slowly.)

**Comment by hcho3** [comment]
@higumachan Yes. We will keep the same C API functions, so you can start building. If you get stuck or don't understand the meaning of an API function, feel free to ask me.

**Pull request merged**